src/Tutor.Core/LearnerModel/DomainOverlay/KnowledgeComponentMasteries/KnowledgeComponentMastery.cs

```diff
@@ -165,9 +128,9 @@ private void TryPass()
             TrySatisfy();
         }
 
-        private void TryComplete(bool isCompletedBeforeSubmission)
+        private void TryComplete()
         {
-            if (isCompletedBeforeSubmission || !IsCompleted) return;
+            if (IsCompleted) return;
 
             Causes(new KnowledgeComponentCompleted
             {
```

Comment:
Shouldn't this method check for completion through the AIMasteries before causing the event?

Reply:
I can't remember what the issue here was. I've changed back IsCompleted to be a field saved in the database as I found it more convenient to work with for the instructor's analytics. Do I still need the isCompletedBeforeSubmission?

Reply:
IsCompletedBeforeSubmission isn't needed, since the IsCompleted property isn't changed until the KCCompleted event is raised now. But TryComplete needs to check that all AIs actually have Submissions before causing the event (so, between line 133 and line 135).